### DISCUSSION TITLE
Alternatives for geospatial dependencies

### DISCUSSION THREAD

**Discussion by rnugent3** [HydrologicEngineeringCenter/HEC-FDA#24] [Q&A] [answered]
@Brennan1994:

"Gonna experiment with using this issue as a discussion board.

Alternative 1: We can continue as is: referencing the specific set of .dll files that we need to run FDA as previously designed. This is already in place and working. It is not upgradable: we don't have source code for these libraries. No capability to fix bugs. Right now these .dlls are just stored in our repo under the lib folder. I could theoretically put these into NuGet Packages and host them on Nexus, which would make this infrastructure similar to alternative 2, except without source code.

Alternative 2: We can upgrade to the latest version of LifeSims Geospatial libraries and package them as NuGet packages. With this method we're basically getting one big upgrade to whatever the latest feature additions/bug fixes are, then we're in a similar spot as before, except that I would reference as NuGet packages. With this path, we could try to keep up with LifeSim's development and I could regularly upgrade our dependencies by pulling latest and generating a new Nugets. It will likely require somewhat regular assistance from Woody to keep everything building, and keep up with API changes.

Alternative 3: We could fork off all LifeSim's  code and maintain it ourselves. We'd then have full control for upgrades and bug fixes, but no assistance from the developer, as they'd no longer be his libraries. We'd be in charge of a lot of code with this alternative we have little no experience with.

Other alternatives would be to drop All or some of our dependencies on LifeSim's stuff, but I think that's a bigger discussion than this issue."

**Comment by rnugent3** [comment]
@rnugent3: 

"It sounds like Alternative 3 is similar to the moves HEC is making generally...bringing all of our code in house so we have control.

Brennan, how comfortable would you be leading the maintenance of this code?"

"FWIW, we also have just more than a little experience with the FDA code base. On the margin, would taking on all that code add a sufficiently large increase in complexity to outweigh the benefits of having control?"

**Comment by rnugent3** [comment]
@HenryGeorgist:

"I don't think forking and maintaining is viable."

**Comment by rnugent3** [comment]
@rnugent3:

"Ok. Which of the alternatives do you think works best?"

**Comment by rnugent3** [comment]
@Brennan1994:

"I'd feel uncomfortable forking and maintaining that codebase. I don't have any familiarity with it, and the owner/maintainer (understandably) would not have any investment in helping with it. I think that would be a hail mary. We'd hope that there are so few bugs/necessary added features with his code that FDA will be dead before we had to make any significant change. It is marginally more secure than alternative 1, because we could make small changes since we have the code, and could sink time into it if we wanted.

I'm afraid Alternative 2 is our safest approach to referencing openGL mapper and lifesim GIS in my opinion. Though I see it being a headache.

Lets move to the discussion board to discuss whether we ought to reference them at all."

**Reply by rnugent3** [reply]
@Brennan1994 I trust your evaluation of the alternatives. 

@HenryGeorgist Do you agree on moving forward with Alternative 2?

**Comment by HenryGeorgist** [comment]
I recommend we move forward with Alternative 1 until coerced. Are there current known bugs or limitations that we are unsatisfied with?

**Reply by Brennan1994** [reply]
The only thing I'm aware of is the BaseMap options which still has an OpenStreetMapping basemap that stopped working. that was replaced with mapbox at some point

I'm willing to sign on to Alternative 1 until it doesn't work for us anymore. It seems like we have awhile before we'll need to be doing map window stuff.

**Comment by rnugent3** [comment]
Alternative 1 it is. Brennan, I agree. I don't think the marginal
contribution of the mapping functionality will be all that big until we can
map the results. Not to say that the QA/QC is not important - but rather
there will be other ways/applications to carry out QA/QC.

**Comment by rnugent3** [comment, marked as answer]
This has been resolved. We'll depend on RAS API.
